File: ADA/Lab_1-search/binary.c
// binary search

# include <stdio.h>
# include <time.h>
# include <stdlib.h>

const int MAXSIZE = 10000;

int RecBinarySearch(int arr[], int low, int high, int target, int *step)
{
	(*step)++;
	int m = (high + low)/2;
	// printf("low = %d \t high = %d \t m = %d\n", arr[low], arr[high], arr[m]);
	if (low > high)
		return -1;
	else
	{
		if (arr[m] == target)
			return m;
		else
		{
			if (target < arr[m])
				// high needs to move up
				return RecBinarySearch(arr, low, m-1, target, step);
			else
				// low need to move down
				return RecBinarySearch(arr, m+1, high, target, step);
		}	
	}
	
}
int main()
{
	int arr[MAXSIZE], target, p, step, n=1;
	FILE *file;
	file = fopen("binary_vals.dat", "w");
	for (n=1; n<=MAXSIZE; n++)
	{
		// get array
		for (int i=0; i<n; i++)
			arr[i] = i*2;
		
		// get random target
		// srand(time(0)+rand());
		// target = arr[rand()%n];

		// get first value of array as target
		target = arr[0];

		// search
		step = 0;
		p = RecBinarySearch(arr, 0, n-1, target, &step);
		fprintf(file, "%d %d\n", n, step);
	}
	int end = time(0);
	fclose(file);
	return 0;
}

File: ADA/Lab_1-search/binary.py
# binary search

import time
from random import randint
import matplotlib.pyplot as plt

def RecBinarySearch(arr, low, high, target, count):
	count += 1
	m = (low + high)//2
	if (low > high):
		return -1, count
	else:
		if arr[m] == target:
			return m, count
		else:
			if target < arr[m]:
				return RecBinarySearch(arr, low, m-1, target, count)
			else:
				return RecBinarySearch(arr, m+1, high, target, count)

def create_arr(size):
	arr = [2*i for i in range(size)]
	return arr

#--------------------------------------

time_list = []
count_list = []
size_list = [i*1000 for i in range(1,500)]

total_start = time.time()
for i in size_list:

	size = i
	arr = create_arr(size)
	# print(f"array is : {arr}")

	# generate random target from array
	target = arr[randint(0,size-1)]

	start = time.time()
	p,count = RecBinarySearch(arr, 0, size-1, target, 0)
	elapsed = time.time() - start
	
	time_list.append(elapsed)
	count_list.append(count)

print(f"{(time.time() - total_start):.3f}sec")
plt.plot(size_list, count_list)
plt.ylim(0,30)
plt.show()


